core/src/main/java/org/infinispan/executors/LimitedExecutor.java

```diff
@@ -0,0 +1,156 @@
+package org.infinispan.executors;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.BiConsumer;
+
+import org.infinispan.util.logging.Log;
+import org.infinispan.util.logging.LogFactory;
+import org.jboss.logging.NDC;
+
+/**
+ * Executes tasks in the given executor, but never has more than {@code maxConcurrentTasks} tasks running at the same
+ * time.
+ *
+ * The number of concurrent tasks can be tweaked at any time with {@link #addPermit()} and {@link #removePermit()}.
+ *
+ * @author Dan Berindei
+ * @since 9.0
+ */
+public class LimitedExecutor implements Executor {
```

Comment:
How about using CompletionService with ThreadPoolExecutor? Because if I understand the idea of this class correctly, we want to execute X tasks concurrently and store results in a list/queue?